### DISCUSSION TITLE
[ideas] Optionally handle method dispatch as dependencies

### DISCUSSION THREAD

**Discussion by multimeric** [ropensci/targets#1300] [Ideas]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Currently functions that are being tracked by targets aren't aware of method dispatch. For example:

```R
foo <- function(x){
  UseMethod("foo")
}

foo.list <- function(){
  "this is a list"
}

list(
  targets::tar_target(
    do_foo,
    foo(list())
  )
)
```

![image](https://github.com/ropensci/targets/assets/5019367/e55e7bf3-9328-4cd6-bdbd-03cabbb1f18b)

In this case, changes to `foo.list` are not going to cause any targets to invalidate, even though it should.

Now, we can't know at pipeline construction time what class `x` will be, but we can in theory pull out all method implementations using `methods`:

```R
> methods(foo)
[1] foo.list
```

Just wondering if there's any place for OOP support such as this in the package.

**Comment by psychelzh** [comment]
The answer is probably no because the package relies on codetools package to do static code analysis and is not able to know which method will be dispatched in such circumstances. See https://github.com/ropensci/targets/discussions/1273#discussioncomment-9241108.

**Reply by wlandau** [reply]
Correct, it is not possible to predict S3 dispatch using static code analysis from `targets`.

**Reply by multimeric** [reply]
Thanks all for the answers. I might look into writing my own target factory that handles this some other way.
